tools/firmware-utils/src/tplink-safeloader.c

```diff
@@ -421,6 +421,45 @@ static struct device_info boards[] = {
 		.last_sysupgrade_partition = "file-system",
 	},
 
+	/** Firmware layout for the C59v2 */
+	{
+		.id	= "ARCHER-C59-V2",
+		.vendor	= "",
+		.support_list =
+			"SupportList:\r\n"
+			"{product_name:Archer C59,product_ver:2.0.0,special_id:00000000}\r\n"
+			"{product_name:Archer C59,product_ver:2.0.0,special_id:55530000}\r\n",
+		.support_trail = '\x00',
+		.soft_ver = "soft_ver:2.0.0 Build 20161206 rel.7303\n",
+
+		.partitions = {
+			{"factory-boot", 0x00000, 0x20000},
+			{"fs-uboot", 0x20000, 0x10000},
+			{"default-mac", 0x30000, 0x00200},
+			{"pin", 0x30200, 0x00200},
+			{"device-id", 0x30400, 0x00100},
+			{"product-info", 0x30500, 0x0fb00},
+			{"os-image", 0x40000, 0x180000},
```

Comment:
if you consider porting to ath79 (would be easy) change this to at least 0x200000 because kernel 4.14 won't fit and you have plenty of space left in file-system

Reply:
I would rather like to see https://patchwork.ozlabs.org/project/openwrt/list/?submitter=73910 tested.

It expects a firmware partition defined instead of os-image and file-system in tplink-safeloader.c. The firmware partition gets splitted into an os-image partition with the size of the actual kernel and the remaining space is used for the file-system. This way we can get rid of the permanent adjustment of the os-image partition if the kernel grows to much.

The kernel and rootfs partitions need to be dropped from the MTDPARTS as well. I'm not entirely sure, but I would suspect they can be already dropped by now - without the linked patchset. Having a firmware partition in the MTDPARTS should be enough to get it working.